### ISSUE TITLE
Docker multi-architecture build issues for DINOv2 API on AWS and Mac

### ISSUE THREAD

**Issue by mdelhaous** [facebookresearch/dinov2#451]
``Background:
I'm working on a FastAPI project that uses Facebook Research's DINOv2 models. The project requires a Docker image with CUDA support for GPU acceleration. The Dockerfile and requirements.txt are as follows:

_**Dockerfile:**_

`# Use the official Python image from the Docker Hub
# FROM python:3.9

# Set the working directory in the container
# WORKDIR /app

FROM nvidia/cuda:11.7.1-cudnn8-runtime-ubuntu20.04

ENV DEBIAN_FRONTEND=noninteractive

# Install Python 3.9 and necessary packages
RUN apt-get update && apt-get install -y \
    python3.9 \
    python3.9-venv \
    python3.9-dev \
    python3-pip \
    && rm -rf /var/lib/apt/lists/*

# Set Python 3.9 as the default Python version
RUN update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.9 1

# Install pip and upgrade it
RUN python3.9 -m pip install --upgrade pip

# Copy the requirements file into the container
COPY requirements.txt .

# Install the dependencies
RUN pip install --no-cache-dir -r requirements.txt

# Copy the rest of the application code into the container
COPY . .

# Set the environment variable to disable xformers
ENV XFORMERS_DISABLED=1

# Expose the port the app runs on
EXPOSE 8000

# Command to run the application
CMD ["uvicorn", "API:app", "--host", "0.0.0.0", "--port", "8000"]
`

_**requirements.txt:**_

`--extra-index-url https://download.pytorch.org/whl/cu117
torch==2.0.0
torchvision==0.15.0
omegaconf
torchmetrics==0.10.3
fvcore
iopath
xformers==0.0.18
submitit
--extra-index-url https://pypi.nvidia.com
cuml-cu11
fastapi
uvicorn
requests
pillow
`
**_API.py:_**
`from fastapi import FastAPI, File, UploadFile, HTTPException, Form
from fastapi.responses import JSONResponse
import torch
from PIL import Image, UnidentifiedImageError
from torchvision import transforms
import io
import requests
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = FastAPI()

MODEL_MAP = {
    'dinov2_vitl14': 'dinov2_vitl14',
    'dinov2_vits14': 'dinov2_vits14',
    'dinov2_vitb14': 'dinov2_vitb14',
    'dinov2_vitg14': 'dinov2_vitg14',
}

def load_model(model_name: str):
    if model_name not in MODEL_MAP:
        raise ValueError(f"Model {model_name} is not supported.")
    model = torch.hub.load('facebookresearch/dinov2', MODEL_MAP[model_name])
    model.eval()
    return model

def preprocess_image(image):
    input_image = Image.open(io.BytesIO(image)).convert('RGB')
    preprocess = transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
    ])
    input_tensor = preprocess(input_image)
    input_batch = input_tensor.unsqueeze(0)
    return input_batch

def infer(model, input_batch):
    with torch.no_grad():
        output = model(input_batch)
    return output

@app.post("/infer/")
async def infer_image(file: UploadFile = File(...), model_name: str = Form(...)):
    try:
        if model_name not in MODEL_MAP:
            raise HTTPException(status_code=400, detail="Invalid model name provided.")

        image_bytes = await file.read()
        input_batch = preprocess_image(image_bytes)
        model = load_model(model_name)
        output = infer(model, input_batch)

        return JSONResponse(content={"inference_output": output.tolist()})
    except Exception as e:
        logger.error(f"Error processing image: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

@app.post("/infer-url/")
async def infer_image_url(url: str = Form(...), model_name: str = Form(...)):
    try:
        if model_name not in MODEL_MAP:
            raise HTTPException(status_code=400, detail="Invalid model name provided.")

        response = requests.get(url)
        if response.status_code != 200:
            logger.error(f"Failed to download image from URL: {url}, Status code: {response.status_code}")
            raise HTTPException(status_code=400, detail="Failed to download image from the provided URL.")

        image_bytes = response.content

        try:
            input_batch = preprocess_image(image_bytes)
        except UnidentifiedImageError:
            logger.error(f"The provided URL does not contain a valid image: {url}")
            raise HTTPException(status_code=400, detail="The provided URL does not contain a valid image.")

        model = load_model(model_name)
        output = infer(model, input_batch)

        return JSONResponse(content={"inference_output": output.tolist()})
    except Exception as e:
        logger.error(f"Error processing image from URL: {url}, Error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
`

Issues:
1. AWS Deployment Error:
When running the Docker container on an AWS machine, I encounter the following error:

**`WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
exec /opt/nvidia/nvidia_entrypoint.sh: exec format error
`**

2. Mac Build Error:
When building the Docker image on a Mac, I encounter the following error:

**`ERROR: Cannot install torch==2.0.1 and torchvision==0.15.1 because these package versions have conflicting dependencies.
ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
`**

**Goal:**
I aim to create multi-architecture Docker builds to ensure compatibility across all platforms (e.g., AMD64, ARM64). I'm not entirely sure about Mac support for DINOv2 yet, but I'm looking for a viable solution.

**Attempts:**

AWS: Tried specifying the platform in the Docker run command, but it didn't resolve the issue.
Mac: Attempted to align dependency versions, but the conflict persists.

**Questions:**

1. How can I resolve the exec format error on AWS?
2. How can I manage the dependency conflict for torch and torchvision during the Docker build on Mac?
3. What are the best practices for creating multi-architecture Docker builds for such projects?

Thank you in advance for your help!
